### DISCUSSION TITLE
usage of typeobject

### DISCUSSION THREAD

**Discussion by xiayh1992** [eProsima/Fast-DDS#3587] [Q&A] [answered]
by read dds doc, it can genarate by fastddsgen -typeobject, but I don't understand typeobject is for what.
DDS layer have five part,  core, doman, publisher, subscriber, topic, I know typeobject is part of topic.  for common usage, I get  PubSubTypes.h as TopicDataType, and then encapsulate it to TypeSupport, then the typesupport can resgister the data type to DomainParticipant by type.register_type(participant),
so, the TypeObject is for what. in the genarated TypeObject file, there are registerXXTypes(std::string type_object_name), the register is to who,  and  what the defference between the the two mentioned register function?

**Comment by xiayh1992** [comment]
I have seen HelloWorldExample and ContentFilteredTopicExample.
In HelloWorldExample, the HelloWorld Constructor,  it just init m_index and m_message.
but in ContentFilteredTopicExample, the HelloWorld Constructor init not only m_index and m_message. but resigterHelloWorldTyples(), what's that for?
someone know the difference of register?

**Comment by xiayh1992** [comment]
14.5. Dynamic Types Discovery and Endpoint Matching, 
the typeObject is firstly mentioned. it mean typeObject is used to Dynamic Types Discovery. but I haven't seen any dynamic type needed by ContentFilteredTopicExample.

**Comment by MiguelCompany** [comment]
@xiayh1992 The type object is used internally when parsing the filter expression, to check it is semantically correct. A DynamicData is also used internally to perform the filtering without fully deserializing the payload.

**Reply by xiayh1992** [reply]
En, It is a wirld step, isn't it?
TopicDataType and TypeSupport have all information about the HelloWorld data, and can register itself to Participant, Why is there need another step to resgister HelloWorld data again?
when scenario of content filter, the participant create filter, and then subscriber use the filter to create data reader. there seem nothing about TypeObject.

**Reply by MiguelCompany** [reply, marked as answer]
> TopicDataType and TypeSupport have all information about the HelloWorld data

This is not true. `TopicDataType` does not have all the information about the `HelloWorld` data. It just knows how to do conversions between a serialized payload and a `HelloWorld` object, but it does not provide methods to allow querying the type of a member given its name, or the offset of a field in the serialized payload.

For the default SQL-like filter, having the TypeObject registered is a requirement, and it is documented as such in [the documentation](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/topic/contentFilteredTopic/createContentFilteredTopic.html): _"Different filter classes may impose different requirements on the related Topic, the expression, or the parameters. The default filter class, in particular, requires that a TypeObject for the related Topic’s type has been registered."_

**Reply by xiayh1992** [reply]
you don't know how I appreciate you!
you mentions the reference in the documentation.
"_Different filter classes may impose different requirements on the related Topic, the expression, or the parameters. The default filter class, in particular, requires that a TypeObject for the related Topic’s type has been registered_."
after the text, the code don't emphasis it, that may comfuse me, I hope the register step can be add in the sample code, it will make usage of contentfilter more clear.
Thanks!

**Reply by MiguelCompany** [reply]
> I hope the register step can be add in the sample code

That's fair. Just opened #3594 for that.
